internal/app/machined/pkg/controllers/network/internal/addressutil/compare.go

```diff
@@ -0,0 +1,89 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package addressutil
+
+import (
+	"cmp"
+	"fmt"
+	"net/netip"
+
+	"github.com/siderolabs/talos/pkg/machinery/nethelpers"
+	"github.com/siderolabs/talos/pkg/machinery/resources/network"
+)
+
+// CompareByAlgorithm returns a comparison function based on the given algorithm.
+func CompareByAlgorithm(algorithm nethelpers.AddressSortAlgorithm) func(a, b netip.Prefix) int {
+	switch algorithm {
+	case nethelpers.AddressSortAlgorithmV1:
+		return ComparePrefixesLegacy
+	case nethelpers.AddressSortAlgorithmV2:
+		return ComparePrefixNew
+	}
+
+	panic(fmt.Sprintf("unknown address sort algorithm: %s", algorithm))
+}
+
+// ComparePrefixesLegacy is the old way to sort prefixes.
+//
+// It only compares addresses and does not take prefix length into account.
+func ComparePrefixesLegacy(a, b netip.Prefix) int {
```

Comment:
this function and the function below is the tl;dr of this change, everything else is just supporting stuff